cmd/goauth/main.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2022 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/artifact-registry-go-tools/pkg/auth"
+)
+
+const help = `
+Handle Go authentication with Google Cloud Artifact Registry Go Repositories.
+
+Add to your GOAUTH environment variable:
+
+  export GOAUTH="sh -c 'GOPROXY=direct go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest <location>'"
```

Comment:
I wonder if we can leave out `sh -c`. Without it the command should still work but not sure if `sh -c` works on Windows?

Reply:
It won't work on Windows and I doubt sh -c is also needed.

```suggestion
  export GOAUTH="go run github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest <location>"
```

must work allright

Reply:
> It won't work on Windows and I doubt sh -c is also needed.
> 
> must work allright

No, it does not. Please feel free to do some testing and verification.

I already checked these scenarios:

```sh
export GOAUTH="go run github.com/smartpricer/artifact-registry-go-tools/cmd/goauth@latest europe-west1"

#--> recursive call

export GOAUTH="GOPROXY=direct go run github.com/smartpricer/artifact-registry-go-tools/cmd/goauth@latest europe-west1"

#--> 401, does not work

export GOAUTH="sh -c 'GOPROXY=direct go run github.com/smartpricer/artifact-registry-go-tools/cmd/goauth@latest europe-west1'"

#--> works fine
```

Reply:
Regarding the Windows support: Feel free to suggest an example for Windows, after testing it.

Reply:
Maybe the instructions must change to ask users to run go install first instead of go run

If you install the binary first like this:

`GOPROXY=direct go install github.com/GoogleCloudPlatform/artifact-registry-go-tools/cmd/goauth@latest`

Then you can use the installed binary directly in GOAUTH without having to use GOPROXY

GOAUTH="goauth europe-west1"

Reply:
Even better would be for this to be distributed with the gcloud SDK like the docker Auth does.

Reply:
I think if going with @cagataygurturk , it needs to be renamed, e.g. `go-auth-google-artifact-registry` or `go-auth-gar` or sth. like this.

Besides that, I kind of agree with @owenhaynes , that this should become part of the Google Cloud SDK.

On our side, we solved this easily, because we run a mono repository, so we already have a place to put team-wide tooling.

Reply:
> Even better would be for this to be distributed with the gcloud SDK like the docker Auth does.

I had the same thought - but the original rationale of implementing auth helpers in the same language as the repository format is so that the users won't have a dependency on gcloud SDK, and some customers do care about this mostly because gcloud is huge.

That being said I think implementing it within gcloud is indeed simpler; I'll go ahead and do that first to hopefully unblock folks. I would like to eventually add support for it in this auth tool, but I think it would require some design review among the team so I'll get back to it later.

Reply:
FYI after spending a lot of time debugging what's described in https://github.com/golang/go/issues/71889 I finally have made it to work locally. Need some extra time adding tests and stuff and this should be available very soon.